daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala

```diff
@@ -0,0 +1,339 @@
+  }
+
+  def relocateA(depth: DepthA, env: Env)(atom: SExprAtomic): SExprAtomic = {
+    makeRelativeA(depth)(makeAbsoluteA(env, atom))
+  }
+
+  type AbsLoc = Either[SELoc, AbsBinding]
+
+  def makeAbsoluteL(env: Env, loc: SELoc): AbsLoc = loc match {
+    case SELocS(rel) => Right(makeAbsoluteB(env, rel))
+    case x: SELocA => Left(x)
+    case x: SELocF => Left(x)
+  }
+
+  def makeRelativeL(depth: DepthA)(loc: AbsLoc): SELoc = loc match {
+    case Left(x: SELocS) => throw CompilationError(s"makeRelativeL: unexpected: $x")
+    case Left(loc) => loc
+    case Right(binding) => SELocS(makeRelativeB(depth, binding))
+  }
+
+  def relocateL(depth: DepthA, env: Env)(loc: SELoc): SELoc = {
+    makeRelativeL(depth)(makeAbsoluteL(env, loc))
+  }
+
+  def flattenExp(depth: DepthA, env: Env, exp: SExpr, k: (AExpr => Trampoline[AExpr])): Res = {
+    Bounce(() => k(transformExp(depth, env, exp, { case (_, sexpr) => Land(AExpr(sexpr)) }).bounce))
+  }
+
+  def transformLet1(depth: DepthA, env: Env, rhs: SExpr, body: SExpr, k: K[SExpr]): Res = {
+    flattenExp(
+      depth,
+      env,
+      rhs, { rhs1 =>
+        val depth1 = DepthA(depth.n + 1)
+        val env1 = trackBindings(depth, env, 1)
+        flattenExp(depth1, env1, body, { body1 =>
+          k(depth, SELet1(rhs1.wrapped, body1.wrapped))
+        })
+      }
+    )
+  }
+
+  /*def transformLet1(depth: DepthA, env: Env, rhs: SExpr, body: SExpr, k: K[SExpr]): Res = {
+    // This is a better transform, but sadly it can blow the stack for deeply nested lets.
+    transformExp(depth, env, rhs, {
+      case (depth, rhs) =>
+        val depth1 = DepthA(depth.n + 1)
+        val env1 = trackBindings(depth, env, 1)
+        val body1 = transformExp(depth1, env1, body, k).bounce
+        Land(SELet1(rhs, body1))
+    })
+  }*/
+
+  def flattenAlts(depth: DepthA, env: Env, alts: Array[SCaseAlt]): Array[SCaseAlt] = {
+    alts.map {
+      case SCaseAlt(pat, body0) =>
+        val n = patternNArgs(pat)
+        val env1 = trackBindings(depth, env, n)
+        SCaseAlt(pat, flattenExp(DepthA(depth.n + n), env1, body0, body => {
+          Land(body)
+        }).bounce.wrapped)
+    }
+  }
+
+  def patternNArgs(pat: SCasePat): Int = pat match {
+    case _: SCPEnum | _: SCPPrimCon | SCPNil | SCPDefault | SCPNone => 0
+    case _: SCPVariant | SCPSome => 1
+    case SCPCons => 2
+  }
+
+  /** `transformExp` is the function at the heart of the ANF transformation.  You can read
+    it's type as saying: "Caller, give me a general expression `exp`, (& depth/env info),
+    and a continuation function `k` which says what you want to do with the transformed
+    expression. Then I will do the transform, and call `k` with it. I reserve the right to
+    wrap further expression-AST around the expression returned by `k`.
+    See: `atomizeExp` for a instance where this wrapping occurs.
+    */
+  def transformExp(depth: DepthA, env: Env, exp: SExpr, k: K[SExpr]): Res =
```

Comment:
Don't we need another parameter for this function which indicates if we really need to produce an atom or if a "simple" application is fine as well? I have a feeling that we currently transform, say,
```haskell
\x -> f x x
```
into
```haskell
\x -> let y = f x x in y
```
Since performance has already significantly improved with the current state, we could do that in a follow-up PR.

Reply:
I created an issue so we don't forget about this: https://github.com/digital-asset/daml/issues/6535

Reply:
Thank you.